Github action run and automatic code references (#6)
flatironinstitute/ccn-template: docs/notes/03-documentation.md

# Documentation

Two decisions are fundamental in this context: the generator and the host. For the generator, we will utilize [MkDocs](https://www.mkdocs.org/), as an alternative to Sphinx. As for the host, we opt for [readthedocs](https://readthedocs.org/) over GitHub Pages. 

> We suggest using [mkdocs-gallery](https://smarie.github.io/mkdocs-gallery/generated/tutorials/plot_parse/#download_links) instead of .ipynb notebooks for tutorials or examples. This is because mkdocs-gallery is easier to review and control versions while still providing the convenience of being downloadable or runnable on Binder as a notebook. These files should reside under the root directory: `examples/` (not `docs/examples/`). However, you can still include `ipynb` files (after clearing the outputs), as they are simpler for contributors to write and pair well with interactive visualization libraries. You can explore both options further [here](https://docs.readthedocs.io/en/stable/guides/jupyter.html) on the readthedocs page.

Reviewer: ```suggestion
We suggest using [mkdocs-gallery](https://smarie.github.io/mkdocs-gallery/generated/tutorials/plot_parse/#download_links) instead of .ipynb notebooks for tutorials or examples. This is because `.py` files ( which `mkdocs-gallery` uses) are easier to review and control versions while still providing the convenience of being downloadable or runnable on Binder as a notebook. These files should reside in a directory within the root directory: `examples/` (not `docs/examples/`). However, you can still include `ipynb` files (after clearing the outputs), as they are simpler for contributors to write and pair well with interactive visualization libraries. Further discussion can be found i the [readthedocs documentation](https://docs.readthedocs.io/en/stable/guides/jupyter.html).
```

# Documentation

Two decisions are fundamental in this context: the generator and the host. For the generator, we will utilize [MkDocs](https://www.mkdocs.org/), as an alternative to Sphinx. As for the host, we opt for [readthedocs](https://readthedocs.org/) over GitHub Pages. 

> We suggest using [mkdocs-gallery](https://smarie.github.io/mkdocs-gallery/generated/tutorials/plot_parse/#download_links) instead of .ipynb notebooks for tutorials or examples. This is because `.py` files ( which `mkdocs-gallery` uses) are easier to review and control versions while still providing the convenience of being downloadable or runnable on Binder as a notebook. These files should reside in a directory within the root directory: `examples/` (not `docs/examples/`). However, you can still include `ipynb` files (after clearing the outputs), as they are simpler for contributors to write and pair well with interactive visualization libraries. Further discussion can be found i the [readthedocs documentation](https://docs.readthedocs.io/en/stable/guides/jupyter.html).

Reviewer: ```suggestion
We suggest using [mkdocs-gallery](https://smarie.github.io/mkdocs-gallery/generated/tutorials/plot_parse/#download_links) instead of .ipynb notebooks for tutorials or examples. This is because `.py` files ( which `mkdocs-gallery` uses) are easier to review and control versions while still providing the convenience of being downloadable or runnable on Binder as a notebook. These files should reside in a directory within the root directory: `examples/` (not `docs/examples/`). However, you can still include `ipynb` files (after clearing the outputs), as they are simpler for contributors to write and pair well with interactive visualization libraries. Further discussion can be found in the [readthedocs documentation](https://docs.readthedocs.io/en/stable/guides/jupyter.html).
```

We suggest using [mkdocs-gallery](https://smarie.github.io/mkdocs-gallery/generated/tutorials/plot_parse/#download_links) instead of .ipynb notebooks for tutorials or examples. This is because mkdocs-gallery is easier to review and control versions while still providing the convenience of being downloadable or runnable on Binder as a notebook. These files should reside under the root directory: `examples/` (not `docs/examples/`). However, you can still include `ipynb` files (after clearing the outputs), as they are simpler for contributors to write and pair well with interactive visualization libraries. You can explore both options further [here](https://docs.readthedocs.io/en/stable/guides/jupyter.html) on the readthedocs page.

Reasoning behind these choices:
- MkDocs supports Markdown out-of-the-box, which is a more straightforward option than ReST.
- readthedocs enables building documentation for PRs, which is highly beneficial.

## Code References

> The [mkdocstrings](https://mkdocstrings.github.io/) plugin for MkDocs allows you to generate the documentation directly from your Python code's docstrings, given they comply with certain standards (refer to the [docstrings](04-docstrings.md) note). 

Author: I think i clarified and suggested the automated for complex projects

Here's the procedure:

1. Ensure your project follows the recommended folder structure.

2. Copy and paste the `docs/gen_ref_pages.py` into your project's `docs/` directory. The script will run each time
mkdocs builds the documentation and will automatically create a `reference/` directory.

3. Confirm that the following plugins are listed in your `mkdocs.yml` file:

```yaml
plugins:
    - search                           # keep the search plugin enabled
    - mkdocstrings                     # plugin for generating documentation from Python docstrings
    - gen-files:
        scripts:
            - docs/gen_ref_pages.py
```

4. Add the **Code Reference** page to your documentation by adding it to the *nav* of the *mkdocs.yml*.

    ```yaml
    nav:
      # other pages in your documentation
      - Code References: reference/ # Note the trailing slash! It is needed so that mkdocs-literate-nav knows 
                                    # it has to look for a SUMMARY.md file in that folder.
    ``` 
    
For a more comprehensive overview of the automatic referencing setup, refer to [this guide](https://mkdocstrings.github.io/recipes/).

**Note on literate navigation:**

> With the "literate-nav" plugin in MkDocs, you can enhance your documentation site's navigation by defining the navigation structure in a separate Markdown file, typically named **SUMMARY.md**. This file essentially serves as a table of contents or a navigation index for your documentation. When the repository is configured with the automatic referencing procedure detailed above, **SUMMARY.md** is temporarily created at build time by the **docs/gen_ref_pages.py** script and is subsequently removed.

Author: I marginally re-edited to match the new structure of the section

### Considerations on Code Referencing

To summarize, automating code referencing offers the following benefits:

1. **Automatic Generation:** There's no need for manual creation of links or copying of code snippets into your 
documentation. The documentation remains up-to-date with each build.

2. **Dynamic Documentation:** As your code and its associated docstrings are updated, the documentation is regenerated.
Changes in script name or code structure do not affect your documentation, which stays up-to-date.

> 3. **Consistent Formatting:** The code reference structure is determined by the **docs/gen_ref_pages.py** script, 

Author: changed as suggested

ensuring consistency.

However, there's a caveat to the automatic generation process: it makes the process of code referencing more 

> challenging to customize. To customize the code references layout, you would need to modify the **gen_ref_pages.py** 

Author: I re-written the whole sentence

3. Make sure that the following plug-ins are listed in your `mkdocs.yml` file:

    ```yaml
    plugins:
        - search                           # make sure the search plugin is still enabled
        - mkdocstrings                     # plugin for generating documentation from Python docstrings
        - gen-files:
            scripts:
                - docs/gen_ref_pages.py     # script for generating reference pages
        - literate-nav:
              nav_file: docs/SUMMARY.md     # navigation file for literate navigation
        - section-index                    # plugin for creating section index
    ```

4. Add the **Code Reference** page to your documentation by adding it to the *nav* of the *mkdocs.yml*.

    ```yaml
    nav:
      # other pages in your documentation
      - Code References: reference/ # Note the trailing slash! It is needed so that mkdocs-literate-nav knows 
                                    # it has to look for a SUMMARY.md file in that folder.
    ``` 
    
For a more detailed description of the automatic referencing setup, see [here](https://mkdocstrings.github.io/recipes/).

**Note on literate navigation:**

> In the context of MkDocs, the "literate-nav" plugin enhances the navigation capabilities of your documentation site by allowing you to define the site's navigation structure in a separate Markdown file (often named SUMMARY.md). This file acts as a table of contents or navigation index for your documentation.

Reviewer: I'm not seeing the `docs/SUMMARY.md`, either on my filesystem (after running `mkdocs serve`) or in the rendered docs -- what am I missing?

Author: I agree I should definitely explain the process of generating doc references. 
I will extend the description to explain what is happening.
The SUMMARY.md is actually temporarily written at build time by the gen_ref_pages.py scripts. 
The fact that many files that would be necessary to render the navigation index and the content of the code references are temporarily created at build time and then removed reduces the clutter of files that would be generated otherwise.

Reviewer: this is clear to me now.

Rationale:
- MkDocs supports markdown out-of-the-box, which is much better than ReST
- readthedocs supports building documentation for PRs, which is very helpful.

## Automatic code references

> Automatic code referencing was setup following the recipe detailed [here](https://mkdocstrings.github.io/recipes/).

Author: it's only one, but they build it step by step

If you were unable to give readthedocs access to your github documentation, you
can use readthedocs' [preview github
action](https://github.com/readthedocs/actions) to post the docs link to the PR.

> ## Generating Documentation From Docstrings 

Reviewer: ```suggestion
## Generating Reference Documentation From Docstrings 
```

> Leverage the [mkdocstrings](https://mkdocstrings.github.io/) plugin for MkDocs to generate documentation directly from your Python code's docstrings. Be sure that your docstrings adhere to the necessary standards. Check the [docstrings](04-docstrings.md) note for reference. With the [autodoc](https://mkdocstrings.github.io/usage/) syntax of `mkdocstrings` you can manually add code documentation to your markdown files for smaller projects.

Reviewer: ```suggestion
Reference documentation consists of technical descriptions of the code and how to use it. This material is rarely read (unlike tutorials or examples), and is instead consulted when users need clarity on specific issues. Fortunately, you've already been writing reference documentation in the form of docstrings. Here, we describe how to automatically construct the reference documentation from these docstrings, using several plugins to `mkdocs`. The following requries that your docstrings adhere to one of the docstring standards, see the [docstrings](04-docstrings.md) note for more details.

Specifically, we will automate the generation of this material using the [mkdocstrings](https://mkdocstrings.github.io/) and [literate-nav](https://mkdocstrings.github.io/) plugins: `mkdocstrings` will generate the docstring documentation and `literate-nav` generates the navigation structure.
```

Reviewer: this also replaces the following paragraph (which starts "For larger...") but figured I'd put it all in one comment

For an in-depth understanding of the following procedure, refer to the `mkdocstrings` [recipes](https://mkdocstrings.github.io/recipes/).

### Procedure

1. Confirm your project follows the recommended [folder structure](01-structure.md).
2. Ensure the following plugins are listed in your `mkdocs.yml` file:

```yaml
plugins:
    - search
    - mkdocstrings
    - gen-files:
        scripts:
            - docs/gen_ref_pages.py
    - literate-nav:
          nav_file: docs/SUMMARY.md
```

3. Make sure the `docs/gen_ref_pages.py` is located in your project's `docs/` directory. This script runs each time mkdocs builds the documentation, automatically creating a `reference/` directory and temporarily creating a `SUMMARY.md` for literate navigation.
4. Add the "Code Reference" page to your documentation by incorporating it into the `nav` of the `mkdocs.yml`:

```yaml
nav:
  # Other pages in your documentation
  - Code References: reference/
``` 

> # Considerations on Code Referencing

Reviewer: ```suggestion
## Why automate?
```

> Automating code referencing offers multiple benefits:

Reviewer: ```suggestion
Automating this procedure offers multiple benefits:
```

1. **Automatic Generation:** Eliminates the need for manual creation of links or copying of code snippets into your documentation. The documentation stays current with each build.
2. **Dynamic Documentation:** Your documentation is regenerated as your code and docstrings are updated. Changes in script names or code structure do not affect your documentation's accuracy.
3. **Consistent Formatting:** The code reference structure is determined by the `docs/gen_ref_pages.py` script, ensuring consistency across the documentation.

> However, automatic generation also presents a caveat: it makes customizing the code referencing process more challenging. If you need to modify the layout of code references, you'll need to alter the **gen_ref_pages.py** script. In contrast, manual referencing offers the flexibility to customize the layout of specific modules by editing the `references.md` file or creating a distinct markdown file for each module and customizing it as needed.

Reviewer: ```suggestion
However, automatic generation also presents a caveat: it makes customizing the code referencing process more challenging. If you need to modify the layout of code references, you'll need to alter the `gen_ref_pages.py` script. In contrast, manual referencing offers the flexibility to customize the layout of specific modules by editing the `references.md` file or creating a distinct markdown file for each module and customizing it as needed. If you wish to do this manually, consult the [mkdocstrings documentation](https://mkdocstrings.github.io/https://mkdocstrings.github.io/).
```

can use readthedocs' [preview github
action](https://github.com/readthedocs/actions) to post the docs link to the PR.

## Generating Reference Documentation From Docstrings 

Reference documentation consists of technical descriptions of the code and how to use it. This material is rarely read (unlike tutorials or examples), and is instead consulted when users need clarity on specific issues. Fortunately, you've already been writing reference documentation in the form of docstrings. Here, we describe how to automatically construct the reference documentation from these docstrings, using several plugins to `mkdocs`. The following requries that your docstrings adhere to one of the docstring standards, see the [docstrings](04-docstrings.md) note for more details.

Specifically, we will automate the generation of this material using the [mkdocstrings](https://mkdocstrings.github.io/) and [literate-nav](https://mkdocstrings.github.io/) plugins: `mkdocstrings` will generate the docstring documentation and `literate-nav` generates the navigation structure.

> For larger, more complex projects, consider automating the code referencing process with the `mkdocstrings` and `literate-nav` plugins. The `literate-nav` plugin lets you define the navigation structure in a separate Markdown file, often named `SUMMARY.md`, enhancing your documentation site's navigation. This file essentially works as a table of contents or a navigation index for your documentation. 

Reviewer: ```suggestion
```

the second paragraph I added covers this material.
